drivers/flash/flash_simulator.c

```diff
@@ -226,8 +231,12 @@ static int flash_sim_write(const struct device *dev, const off_t offset,
 
 	FLASH_SIM_STATS_INC(flash_sim_stats, flash_write_calls);
 
+#if defined(CONFIG_FLASH_SIMULATOR_EXPLICIT_ERASE)
 	/* check if any unit has been already programmed */
 	memset(buf, FLASH_SIMULATOR_ERASE_VALUE, sizeof(buf));
+#else
+	memcpy(buf, MOCK_FLASH(offset), sizeof(buf));
```

Comment:
This doesn't seem right. You fill `buf` with current bytes [0, FLASH_SIMULATOR_PROG_UNIT) and then compare it to all [FLASH_SIMULATOR_PROG_UNIT * n, FLASH_SIMULATOR_PROG_UNIT  *(n+1)) to detect double writes. Shouldn't you rather skip the whole for loop if there's no explicit erase?